// @clickhouse/client/dist/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConnection = exports.createClient = void 0;
const client_common_1 = require("@clickhouse/client-common");
const connection_1 = require("./connection");
const result_set_1 = require("./result_set");
const utils_1 = require("./utils");
function createClient(config) {
    var _a, _b, _c, _d, _e, _f;
    let tls = undefined;
    if (config === null || config === void 0 ? void 0 : config.tls) {
        if ('cert' in config.tls && 'key' in config.tls) {
            tls = {
                type: 'Mutual',
                ...config.tls,
            };
        }
        else {
            tls = {
                type: 'Basic',
                ...config.tls,
            };
        }
    }
    const keep_alive = {
        enabled: (_b = (_a = config === null || config === void 0 ? void 0 : config.keep_alive) === null || _a === void 0 ? void 0 : _a.enabled) !== null && _b !== void 0 ? _b : true,
        socket_ttl: (_d = (_c = config === null || config === void 0 ? void 0 : config.keep_alive) === null || _c === void 0 ? void 0 : _c.socket_ttl) !== null && _d !== void 0 ? _d : 2500,
        retry_on_expired_socket: (_f = (_e = config === null || config === void 0 ? void 0 : config.keep_alive) === null || _e === void 0 ? void 0 : _e.retry_on_expired_socket) !== null && _f !== void 0 ? _f : false,
    };
    return new client_common_1.ClickHouseClient({
        impl: {
            make_connection: (params) => {
                switch (params.url.protocol) {
                    case 'http:':
                        return new connection_1.NodeHttpConnection({ ...params, keep_alive });
                    case 'https:':
                        return new connection_1.NodeHttpsConnection({ ...params, tls, keep_alive });
                    default:
                        throw new Error('Only HTTP(s) adapters are supported');
                }
            },
            make_result_set: (stream, format, session_id) => new result_set_1.ResultSet(stream, format, session_id),
            values_encoder: new utils_1.NodeValuesEncoder(),
            close_stream: async (stream) => {
                stream.destroy();
            },
        },
        ...(config || {}),
    });
}
exports.createClient = createClient;
function createConnection(params) {
    // TODO throw ClickHouseClient error
    switch (params.url.protocol) {
        case 'http:':
            return new connection_1.NodeHttpConnection(params);
        case 'https:':
            return new connection_1.NodeHttpsConnection(params);
        default:
            throw new Error('Only HTTP(s) adapters are supported');
    }
}
exports.createConnection = createConnection;
