// @clickhouse/client/dist/connection/node_https_connection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeHttpsConnection = void 0;
const client_common_1 = require("@clickhouse/client-common");
const https_1 = __importDefault(require("https"));
const node_base_connection_1 = require("./node_base_connection");
class NodeHttpsConnection extends node_base_connection_1.NodeBaseConnection {
    constructor(params) {
        var _a, _b, _c;
        const agent = new https_1.default.Agent({
            keepAlive: params.keep_alive.enabled,
            maxSockets: params.max_open_connections,
            ca: (_a = params.tls) === null || _a === void 0 ? void 0 : _a.ca_cert,
            key: ((_b = params.tls) === null || _b === void 0 ? void 0 : _b.type) === 'Mutual' ? params.tls.key : undefined,
            cert: ((_c = params.tls) === null || _c === void 0 ? void 0 : _c.type) === 'Mutual' ? params.tls.cert : undefined,
        });
        super(params, agent);
    }
    buildDefaultHeaders(username, password) {
        var _a, _b;
        if (((_a = this.params.tls) === null || _a === void 0 ? void 0 : _a.type) === 'Mutual') {
            return {
                'X-ClickHouse-User': username,
                'X-ClickHouse-Key': password,
                'X-ClickHouse-SSL-Certificate-Auth': 'on',
            };
        }
        if (((_b = this.params.tls) === null || _b === void 0 ? void 0 : _b.type) === 'Basic') {
            return {
                'X-ClickHouse-User': username,
                'X-ClickHouse-Key': password,
            };
        }
        return super.buildDefaultHeaders(username, password);
    }
    createClientRequest(params) {
        return https_1.default.request(params.url, {
            method: params.method,
            agent: this.agent,
            headers: (0, client_common_1.withCompressionHeaders)({
                headers: this.headers,
                compress_request: params.compress_request,
                decompress_response: params.decompress_response,
            }),
            signal: params.abort_signal,
        });
    }
}
exports.NodeHttpsConnection = NodeHttpsConnection;
