// @clickhouse/client-common/dist/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClickHouseClient = void 0;
const client_common_1 = require("@clickhouse/client-common");
function validateConnectionParams({ url }) {
    if (url.protocol !== 'http:' && url.protocol !== 'https:') {
        throw new Error(`Only http(s) protocol is supported, but given: [${url.protocol}]`);
    }
}
function createUrl(host) {
    try {
        return new URL(host);
    }
    catch (err) {
        throw new Error('Configuration parameter "host" contains malformed url.');
    }
}
function getConnectionParams(config) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    return {
        application_id: config.application,
        url: createUrl((_a = config.host) !== null && _a !== void 0 ? _a : 'http://localhost:8123'),
        request_timeout: (_b = config.request_timeout) !== null && _b !== void 0 ? _b : 300000,
        max_open_connections: (_c = config.max_open_connections) !== null && _c !== void 0 ? _c : Infinity,
        compression: {
            decompress_response: (_e = (_d = config.compression) === null || _d === void 0 ? void 0 : _d.response) !== null && _e !== void 0 ? _e : true,
            compress_request: (_g = (_f = config.compression) === null || _f === void 0 ? void 0 : _f.request) !== null && _g !== void 0 ? _g : false,
        },
        username: (_h = config.username) !== null && _h !== void 0 ? _h : 'default',
        password: (_j = config.password) !== null && _j !== void 0 ? _j : '',
        database: (_k = config.database) !== null && _k !== void 0 ? _k : 'default',
        clickhouse_settings: (_l = config.clickhouse_settings) !== null && _l !== void 0 ? _l : {},
        logWriter: new client_common_1.LogWriter(((_m = config === null || config === void 0 ? void 0 : config.log) === null || _m === void 0 ? void 0 : _m.LoggerClass)
            ? new config.log.LoggerClass()
            : new client_common_1.DefaultLogger(), (_o = config.log) === null || _o === void 0 ? void 0 : _o.level),
    };
}
class ClickHouseClient {
    constructor(config) {
        Object.defineProperty(this, "connectionParams", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "connection", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "makeResultSet", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "valuesEncoder", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "closeStream", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "sessionId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.connectionParams = getConnectionParams(config);
        this.sessionId = config.session_id;
        validateConnectionParams(this.connectionParams);
        this.connection = config.impl.make_connection(this.connectionParams);
        this.makeResultSet = config.impl.make_result_set;
        this.valuesEncoder = config.impl.values_encoder;
        this.closeStream = config.impl.close_stream;
    }
    getQueryParams(params) {
        return {
            clickhouse_settings: {
                ...this.connectionParams.clickhouse_settings,
                ...params.clickhouse_settings,
            },
            query_params: params.query_params,
            abort_signal: params.abort_signal,
            query_id: params.query_id,
            session_id: this.sessionId,
        };
    }
    /**
     * Used for most statements that can have a response, such as SELECT.
     * FORMAT clause should be specified separately via {@link QueryParams.format} (default is JSON)
     * Consider using {@link ClickHouseClient.insert} for data insertion,
     * or {@link ClickHouseClient.command} for DDLs.
     */
    async query(params) {
        var _a;
        const format = (_a = params.format) !== null && _a !== void 0 ? _a : 'JSON';
        const query = formatQuery(params.query, format);
        const { stream, query_id } = await this.connection.query({
            query,
            ...this.getQueryParams(params),
        });
        return this.makeResultSet(stream, format, query_id);
    }
    /**
     * It should be used for statements that do not have any output,
     * when the format clause is not applicable, or when you are not interested in the response at all.
     * Response stream is destroyed immediately as we do not expect useful information there.
     * Examples of such statements are DDLs or custom inserts.
     * If you are interested in the response data, consider using {@link ClickHouseClient.exec}
     */
    async command(params) {
        const { stream, query_id } = await this.exec(params);
        await this.closeStream(stream);
        return { query_id };
    }
    /**
     * Similar to {@link ClickHouseClient.command}, but for the cases where the output is expected,
     * but format clause is not applicable. The caller of this method is expected to consume the stream,
     * otherwise, the request will eventually be timed out.
     */
    async exec(params) {
        const query = removeTrailingSemi(params.query.trim());
        return await this.connection.exec({
            query,
            ...this.getQueryParams(params),
        });
    }
    /**
     * The primary method for data insertion. It is recommended to avoid arrays in case of large inserts
     * to reduce application memory consumption and consider streaming for most of such use cases.
     * As the insert operation does not provide any output, the response stream is immediately destroyed.
     * In case of a custom insert operation, such as, for example, INSERT FROM SELECT,
     * consider using {@link ClickHouseClient.command}, passing the entire raw query there (including FORMAT clause).
     */
    async insert(params) {
        const format = params.format || 'JSONCompactEachRow';
        this.valuesEncoder.validateInsertValues(params.values, format);
        const query = `INSERT INTO ${params.table.trim()} FORMAT ${format}`;
        return await this.connection.insert({
            query,
            values: this.valuesEncoder.encodeValues(params.values, format),
            ...this.getQueryParams(params),
        });
    }
    /**
     * Health-check request. Can throw an error if the connection is refused.
     */
    async ping() {
        return await this.connection.ping();
    }
    /**
     * Shuts down the underlying connection.
     * This method should ideally be called only once per application lifecycle,
     * for example, during the graceful shutdown phase.
     */
    async close() {
        return await this.connection.close();
    }
}
exports.ClickHouseClient = ClickHouseClient;
function formatQuery(query, format) {
    query = query.trim();
    query = removeTrailingSemi(query);
    return query + ' \nFORMAT ' + format;
}
function removeTrailingSemi(query) {
    let lastNonSemiIdx = query.length;
    for (let i = lastNonSemiIdx; i > 0; i--) {
        if (query[i - 1] !== ';') {
            lastNonSemiIdx = i;
            break;
        }
    }
    if (lastNonSemiIdx !== query.length) {
        return query.slice(0, lastNonSemiIdx);
    }
    return query;
}
